### PR TITLE
[WIP][RFC] [Do not review - still jamming, still lots of bugs and hacks] Serializable GuardedCode for cross-process dynamo caching

### PR DESCRIPTION
**Pull request by voznesenskym**
Test script:
```
class CustomLayer(nn.Module):
    def __init__(self, size):
        super(CustomLayer, self).__init__()
        # Initialize a constant tensor for element-wise multiplication
        self.weights = nn.Parameter(torch.randn(size), requires_grad=True)

    def forward(self, x):
        # Element-wise multiplication
        return x * self.weights

class SampleNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(SampleNN, self).__init__()
        self.custom_layer = CustomLayer(input_size)
        self.linear1 = nn.Linear(input_size, hidden_size)
        self.linear2 = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.custom_layer(x)
        x = F.relu(self.linear1(x))
        x = self.linear2(x)
        return x

input_size = 10
hidden_size = 20
output_size = 5

sample_nn = SampleNN(input_size, hidden_size, output_size)

def my_backend(gm, inputs):
    print(gm)
    # breakpoint()
    return gm

nn = torch._dynamo.optimize(my_backend, remote=True)(sample_nn)
nn(torch.randn([1, input_size]))
nn(torch.randn([1, input_size]))
nn(torch.randn([1, input_size]))
exit(0)
```

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

**Comment by voznesenskym** [comment]
I am still iterating, but the test example works. Still very trivial, still a lot of directional changes to make. One major gap in that PR, is that the flag in the config will 100% not ship, we want to collapse the two worlds.

**Comment by voznesenskym** [comment]
This works with graph breaks now:

```
class CustomLayer(nn.Module):
    def __init__(self, size):
        super(CustomLayer, self).__init__()
        # Initialize a constant tensor for element-wise multiplication
        self.weights = nn.Parameter(torch.randn(size), requires_grad=True)

    def forward(self, x):
        # Element-wise multiplication
        return x * self.weights

class SampleNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(SampleNN, self).__init__()
        self.custom_layer = CustomLayer(input_size)
        self.linear1 = nn.Linear(input_size, hidden_size)
        self.linear2 = nn.Linear(hidden_size, output_size)

    def forward(self, x):
        x = self.custom_layer(x)
        x = F.relu(self.linear1(x))
        x = self.linear2(x) * 2
        print("woowoo")
        x = x * x
        return x

input_size = 10
hidden_size = 20
output_size = 5

sample_nn = SampleNN(input_size, hidden_size, output_size)

def my_backend(gm, inputs):
    print(gm)
    return gm

nn = torch._dynamo.optimize(my_backend, serialize=True)(sample_nn)
nn(torch.randn([1, input_size]))
nn(torch.randn([1, input_size]))
nn(torch.randn([1, input_size]))
```

**Comment by voznesenskym** [comment]
Long term steps for serde in this PR:

1) P0: Write to disk, both real guards and serialization guards
2) P1: Load from disk, run serialization guards
3) P1: (2) passed - we install the real guards, not the serialization guards

Right now in the prototype, I have a hack where step 3 installs the serialization guards, cause those suffice to unblock. However, in the real fix, we will seed the serialized code with sources/hints and enough info to bootstrap the guards and create the fastpath/runtime guards.
